File: app/src/main/java/org/alkaline/taskbrain/dsl/builtins/ListFunctions.kt
package org.alkaline.taskbrain.dsl.builtins

import org.alkaline.taskbrain.dsl.runtime.Arguments
import org.alkaline.taskbrain.dsl.runtime.values.BooleanVal
import org.alkaline.taskbrain.dsl.runtime.BuiltinFunction
import org.alkaline.taskbrain.dsl.runtime.BuiltinRegistry
import org.alkaline.taskbrain.dsl.runtime.values.DateTimeVal
import org.alkaline.taskbrain.dsl.runtime.values.DateVal
import org.alkaline.taskbrain.dsl.runtime.values.DslValue
import org.alkaline.taskbrain.dsl.runtime.Environment
import org.alkaline.taskbrain.dsl.runtime.ExecutionException
import org.alkaline.taskbrain.dsl.runtime.values.LambdaVal
import org.alkaline.taskbrain.dsl.runtime.values.ListVal
import org.alkaline.taskbrain.dsl.runtime.values.NoteVal
import org.alkaline.taskbrain.dsl.runtime.values.NumberVal
import org.alkaline.taskbrain.dsl.runtime.values.StringVal
import org.alkaline.taskbrain.dsl.runtime.values.TimeVal
import org.alkaline.taskbrain.dsl.runtime.values.UndefinedVal

/**
 * List-related builtin functions.
 *
 * list(), sort(), first()
 */
object ListFunctions {

    fun register(registry: BuiltinRegistry) {
        registry.register(listFunction)
        registry.register(sortFunction)
        registry.register(firstFunction)
    }

    /**
     * list(a, b, c, ...) - Create a list from arguments.
     *
     * Parameters:
     * - Any number of positional arguments (including zero)
     *
     * Returns: A ListVal containing all the arguments
     *
     * Examples:
     *   [list()]              - Empty list
     *   [list(1, 2, 3)]       - List of numbers
     *   [list("a", "b")]      - List of strings
     *   [list(1, "mixed", date)]  - Mixed types allowed
     */
    private val listFunction = BuiltinFunction(
        name = "list",
        isDynamic = false
    ) { args, _ ->
        ListVal(args.positional)
    }

    /**
     * sort(list, key: lambda, order: ascending|descending) - Sort a list of values.
     *
     * Parameters:
     * - First positional argument: The list to sort
     * - key: Optional lambda to extract sort key from each item
     * - order: Optional `ascending` (default) or `descending` for sort direction
     *
     * Returns: A new sorted ListVal
     *
     * Comparison rules:
     * - Numbers: numerically
     * - Strings: lexicographically
     * - Dates/Times/DateTimes: chronologically
     * - Booleans: false < true
     * - Notes: by path, then by name, then by id
     * - Other types: by type name then display string
     *
     * Examples:
     *   [sort(find())]                                      - Sort notes by default (path/name/id)
     *   [sort(find(), key: [i.path])]                 - Sort by path
     *   [sort(find(), key: [i.path], order: descending)]  - Sort by path descending
     *   [sort(list(3, 1, 4), order: descending)]            - Sort numbers descending
     */
    private val sortFunction = BuiltinFunction(
        name = "sort",
        isDynamic = false  // Results are deterministic based on input
    ) { args, env ->
        // Require exactly one positional argument (the list)
        if (args.size != 1) {
            throw ExecutionException("'sort' requires exactly 1 positional argument (list), got ${args.size}")
        }

        val listArg = args[0]
            ?: throw ExecutionException("'sort' requires a list argument")
        val list = listArg as? ListVal
            ?: throw ExecutionException("'sort' first argument must be a list, got ${listArg.typeName}")

        val keyLambda = args.getLambda("key")
        val orderArg = args["order"]
        val isDescending = when (orderArg) {
            null -> false  // Default: ascending
            is StringVal -> {
                val value = orderArg.value.lowercase()
                when (value) {
                    SortConstants.ASCENDING -> false
                    SortConstants.DESCENDING -> true
                    else -> throw ExecutionException(
                        "'sort' order must be 'ascending' or 'descending', got '$value'"
                    )
                }
            }
            else -> throw ExecutionException(
                "'sort' order must be 'ascending' or 'descending', got ${orderArg.typeName}"
            )
        }

        val sorted = sortList(list.items, keyLambda, env)
        ListVal(if (isDescending) sorted.reversed() else sorted)
    }

    /**
     * Sort a list of DslValues using optional key lambda.
     */
    private fun sortList(
        items: List<DslValue>,
        keyLambda: LambdaVal?,
        env: Environment
    ): List<DslValue> {
        if (items.isEmpty()) return items

        return if (keyLambda != null) {
            val executor = env.getExecutor()
                ?: throw ExecutionException("'sort' key: requires an executor in the environment")

            // Extract keys once and sort by them
            val itemsWithKeys = items.map { item ->
                item to executor.invokeLambda(keyLambda, listOf(item))
            }
            itemsWithKeys.sortedWith { (_, key1), (_, key2) ->
                compareValues(key1, key2)
            }.map { it.first }
        } else {
            // Sort by the items themselves
            items.sortedWith { a, b -> compareValues(a, b) }
        }
    }

    /**
     * Compare two DslValues for sorting.
     *
     * Comparison rules:
     * 1. Same type: compare by value semantics
     * 2. Different types: compare by type precedence
     *
     * Type precedence (lowest to highest):
     * undefined < boolean < number < string < date < time < datetime < note < list < other
     */
    internal fun compareValues(a: DslValue, b: DslValue): Int {
        // Handle undefined - always sorts first
        if (a is UndefinedVal && b is UndefinedVal) return 0
        if (a is UndefinedVal) return -1
        if (b is UndefinedVal) return 1

        // Same type comparison
        return when {
            a is NumberVal && b is NumberVal -> a.value.compareTo(b.value)
            a is StringVal && b is StringVal -> a.value.compareTo(b.value)
            a is BooleanVal && b is BooleanVal -> a.value.compareTo(b.value)
            a is DateVal && b is DateVal -> a.value.compareTo(b.value)
            a is TimeVal && b is TimeVal -> a.value.compareTo(b.value)
            a is DateTimeVal && b is DateTimeVal -> a.value.compareTo(b.value)
            a is NoteVal && b is NoteVal -> compareNotes(a, b)
            a is ListVal && b is ListVal -> compareLists(a, b)

            // Different types: compare by type precedence
            else -> typePrecedence(a).compareTo(typePrecedence(b))
        }
    }

    /**
     * Compare two notes for sorting.
     * Order: by path, then by name (first line), then by id.
     */
    private fun compareNotes(a: NoteVal, b: NoteVal): Int {
        // First compare by path
        val pathCompare = a.note.path.compareTo(b.note.path)
        if (pathCompare != 0) return pathCompare

        // Then by name (first line of content)
        val nameA = a.note.content.lines().firstOrNull() ?: ""
        val nameB = b.note.content.lines().firstOrNull() ?: ""
        val nameCompare = nameA.compareTo(nameB)
        if (nameCompare != 0) return nameCompare

        // Finally by id
        return a.note.id.compareTo(b.note.id)
    }

    /**
     * Compare two lists for sorting.
     * Order: by size first, then element-by-element.
     */
    private fun compareLists(a: ListVal, b: ListVal): Int {
        // First compare by size
        val sizeCompare = a.size.compareTo(b.size)
        if (sizeCompare != 0) return sizeCompare

        // Then compare element by element
        for (i in 0 until a.size) {
            val elemCompare = compareValues(a.items[i], b.items[i])
            if (elemCompare != 0) return elemCompare
        }

        return 0
    }

    /**
     * Get type precedence for cross-type comparison.
     */
    private fun typePrecedence(value: DslValue): Int = when (value) {
        is UndefinedVal -> 0
        is BooleanVal -> 1
        is NumberVal -> 2
        is StringVal -> 3
        is DateVal -> 4
        is TimeVal -> 5
        is DateTimeVal -> 6
        is NoteVal -> 7
        is ListVal -> 8
        else -> 9  // Lambda, Pattern, etc.
    }

    /**
     * first(list) - Get the first item from a list.
     *
     * Returns: The first item, or UndefinedVal if the list is empty.
     *
     * Following the Mindl design principle: graceful undefined access.
     * Use `maybe(first(...))` to convert undefined to empty if needed.
     *
     * Examples:
     *   [first(find(path: "inbox"))]  - Get the inbox note if it exists
     *   [first(sort(find(), order: desc))]  - Get the most recent note
     */
    private val firstFunction = BuiltinFunction(
        name = "first",
        isDynamic = false
    ) { args, _ ->
        if (args.size != 1) {
            throw ExecutionException("'first' requires exactly 1 argument (list), got ${args.size}")
        }

        val listArg = args[0]
            ?: throw ExecutionException("'first' requires a list argument")
        val list = listArg as? ListVal
            ?: throw ExecutionException("'first' argument must be a list, got ${listArg.typeName}")

        // Return first item or undefined if empty (graceful undefined access)
        list.items.firstOrNull() ?: UndefinedVal
    }
}

File: web/src/dsl/builtins/ListFunctions.ts
import type { DslValue, LambdaVal } from '../runtime/DslValue'
import { listVal, UNDEFINED } from '../runtime/DslValue'
import type { BuiltinFunction } from '../runtime/BuiltinRegistry'
import { ExecutionException } from '../runtime/ExecutionException'
import type { Environment } from '../runtime/Environment'
import { ASCENDING, DESCENDING } from './SortConstants'

export function getListFunctions(): BuiltinFunction[] {
  return [listFunction, sortFunction, firstFunction]
}

const listFunction: BuiltinFunction = {
  name: 'list',
  call: (args) => listVal(args.positional),
}

const sortFunction: BuiltinFunction = {
  name: 'sort',
  call: (args, env) => {
    args.assertNoUnknownNamed('sort', 'key', 'order')
    if (args.size !== 1) throw new ExecutionException(`'sort' requires exactly 1 positional argument (list), got ${args.size}`)
    const listArg = args.getPositional(0)
    if (!listArg) throw new ExecutionException("'sort' requires a list argument")
    if (listArg.kind !== 'ListVal') throw new ExecutionException(`'sort' first argument must be a list, got ${listArg.kind}`)

    const keyLambda = args.getLambda('key')
    const orderArg = args.getNamed('order')
    let isDescending = false
    if (orderArg) {
      if (orderArg.kind !== 'StringVal') throw new ExecutionException(`'sort' order must be 'ascending' or 'descending', got ${orderArg.kind}`)
      const value = orderArg.value.toLowerCase()
      if (value === ASCENDING) isDescending = false
      else if (value === DESCENDING) isDescending = true
      else throw new ExecutionException(`'sort' order must be 'ascending' or 'descending', got '${value}'`)
    }

    const sorted = sortList(listArg.items, keyLambda, env)
    return listVal(isDescending ? sorted.reverse() : sorted)
  },
}

const firstFunction: BuiltinFunction = {
  name: 'first',
  call: (args) => {
    if (args.size !== 1) throw new ExecutionException(`'first' requires exactly 1 argument (list), got ${args.size}`)
    const listArg = args.getPositional(0)
    if (!listArg) throw new ExecutionException("'first' requires a list argument")
    if (listArg.kind !== 'ListVal') throw new ExecutionException(`'first' argument must be a list, got ${listArg.kind}`)
    return listArg.items[0] ?? UNDEFINED
  },
}

function sortList(items: DslValue[], keyLambda: LambdaVal | null, env: Environment): DslValue[] {
  if (items.length === 0) return items

  if (keyLambda) {
    const executor = env.getExecutor()
    if (!executor) throw new ExecutionException("'sort' key: requires an executor in the environment")
    const itemsWithKeys = items.map((item) => ({
      item,
      key: executor.invokeLambda(keyLambda, [item]),
    }))
    itemsWithKeys.sort((a, b) => sortCompareValues(a.key, b.key))
    return itemsWithKeys.map((x) => x.item)
  }

  return [...items].sort(sortCompareValues)
}

/**
 * Compare two DslValues for sorting with type precedence.
 */
export function sortCompareValues(a: DslValue, b: DslValue): number {
  if (a.kind === 'UndefinedVal' && b.kind === 'UndefinedVal') return 0
  if (a.kind === 'UndefinedVal') return -1
  if (b.kind === 'UndefinedVal') return 1

  if (a.kind === 'NumberVal' && b.kind === 'NumberVal') return a.value - b.value
  if (a.kind === 'StringVal' && b.kind === 'StringVal') return a.value < b.value ? -1 : a.value > b.value ? 1 : 0
  if (a.kind === 'BooleanVal' && b.kind === 'BooleanVal') return Number(a.value) - Number(b.value)
  if (a.kind === 'DateVal' && b.kind === 'DateVal') return a.value < b.value ? -1 : a.value > b.value ? 1 : 0
  if (a.kind === 'TimeVal' && b.kind === 'TimeVal') return a.value < b.value ? -1 : a.value > b.value ? 1 : 0
  if (a.kind === 'DateTimeVal' && b.kind === 'DateTimeVal') return a.value < b.value ? -1 : a.value > b.value ? 1 : 0
  if (a.kind === 'NoteVal' && b.kind === 'NoteVal') return compareNotes(a, b)
  if (a.kind === 'ListVal' && b.kind === 'ListVal') return compareLists(a, b)

  return typePrecedence(a) - typePrecedence(b)
}

function compareNotes(a: DslValue & { kind: 'NoteVal' }, b: DslValue & { kind: 'NoteVal' }): number {
  const pathCmp = a.note.path < b.note.path ? -1 : a.note.path > b.note.path ? 1 : 0
  if (pathCmp !== 0) return pathCmp
  const nameA = a.note.content.split('\n')[0] ?? ''
  const nameB = b.note.content.split('\n')[0] ?? ''
  const nameCmp = nameA < nameB ? -1 : nameA > nameB ? 1 : 0
  if (nameCmp !== 0) return nameCmp
  return a.note.id < b.note.id ? -1 : a.note.id > b.note.id ? 1 : 0
}

function compareLists(a: DslValue & { kind: 'ListVal' }, b: DslValue & { kind: 'ListVal' }): number {
  const sizeCmp = a.items.length - b.items.length
  if (sizeCmp !== 0) return sizeCmp
  for (let i = 0; i < a.items.length; i++) {
    const cmp = sortCompareValues(a.items[i]!, b.items[i]!)
    if (cmp !== 0) return cmp
  }
  return 0
}

function typePrecedence(value: DslValue): number {
  switch (value.kind) {
    case 'UndefinedVal': return 0
    case 'BooleanVal': return 1
    case 'NumberVal': return 2
    case 'StringVal': return 3
    case 'DateVal': return 4
    case 'TimeVal': return 5
    case 'DateTimeVal': return 6
    case 'NoteVal': return 7
    case 'ListVal': return 8
    default: return 9
  }
}
